### PR TITLE
Update module documentation

### PR DESCRIPTION
**Pull request by SergioGasquez**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
First PR for unifying the format of module documentation. Used this format:
```
//! # Peripheral Name (Peripheral Acronym)
//!
//! ## Overview
//! Small description of the peripheral, see ESP-IDF docs or TRM 
//! 
//! ## Configuration
//! Explain how can the peripheral be configured, and which parameters can be configured
//! 
//! ## Usage
//! Explain if we implement any external traits
//! 
//! ## Examples
//! ### Name of the Example
//! Small description of the example if needed
//! ```rust, no_run
//! ...
//! ```
//! ## Implementation State
//! Missing features...
```

Still a lot of modules to be updated, but this is already a big PR (sorry)

Should close https://github.com/esp-rs/esp-hal/issues/1660

#### Testing
Built the documentation and ran the doc-tests.

### CODE REVIEW

**Comment by jessebraham** [comment]
Thanks for this. Will probably take me a couple days to review with that many files changes, can't see myself sitting down and just powering through all that, just FYI.

File: esp-hal/src/soc/esp32/gpio.rs [hunk 1]
```diff
@@ -61,11 +61,16 @@ pub const NUM_PINS: usize = 40;
 
 pub(crate) const FUNC_IN_SEL_OFFSET: usize = 0;
 
+#[doc(hidden)]
```
Comment (jessebraham) [inline esp-hal/src/soc/esp32/gpio.rs:64]: This applies to all changes in the `soc/$CHIP/gpio.rs` file: do these types actually need to be public? Can they be `pub(crate)` instead? I would really prefer not to litter too many `#[doc(hidden)]` attributes throughout the code base, it usually implies something is not quite correct.

Comment (SergioGasquez) [reply esp-hal/src/soc/esp32/gpio.rs]: Updated!

File: esp-hal/src/aes/mod.rs [hunk 2]
```diff
@@ -1,17 +1,21 @@
-//! # Advanced Encryption Standard (AES) support.
+//! # Advanced Encryption Standard (AES).
 //!
 //! ## Overview
+//! The AES accelerator is a a hardware device that speeds up computation
```
Comment (JurajSadel) [inline esp-hal/src/aes/mod.rs:4]:
```suggestion
//! The AES accelerator is a hardware device that speeds up computation
```

File: esp-hal/src/analog/adc/mod.rs [hunk 3]
```diff
@@ -1,11 +1,24 @@
 //! # Analog to Digital Converter (ADC)
 //!
-//! The Analog to Digital Converter (ADC) is integrated on the chip, and is
-//! capable of measuring analog signals from specific analog I/O pins. One or
-//! more ADC units are available, depending on the device being used.
+//! ## Overview
+//! The ADC is integrated on the chip, and is capable of measuring analog
```
Comment (JurajSadel) [inline esp-hal/src/analog/adc/mod.rs:4]: Maybe we can leave the Analog to Digital Converter here?

File: esp-hal/src/analog/dac.rs [hunk 4]
```diff
@@ -1,16 +1,20 @@
 //! # Digital to Analog Converter (DAC)
 //!
-//! The `dac` module enables users to generate analog output signals with
-//! precise control over voltage levels using one of the onboard
-//! digital-to-analog converters (DAC).
+//! ## Overview
+//! Espressif devices usually have multiple DAC channels. Each DAC channel can
```
Comment (JurajSadel) [inline esp-hal/src/analog/dac.rs:4]: Same as above?

File: esp-hal/src/interrupt/mod.rs [hunk 5]
```diff
@@ -1,14 +1,31 @@
 //! # Interrupt support
 //!
+//! ## Overview
+//! This modules routes one or multiple peripheral interrupt sources to any one
```
Comment (JurajSadel) [inline esp-hal/src/interrupt/mod.rs:4]:
```suggestion
//! This module routes one or multiple peripheral interrupt sources to any one
```

File: esp-hal/src/pcnt/mod.rs [hunk 6]
```diff
@@ -1,40 +1,24 @@
...
 //!
 //! ## Overview
-//! The `PCNT (Pulse Counter)` driver for `ESP` chips is a software component
-//! that provides an interface for controlling and utilizing the `PCNT`
-//! peripheral. The `PCNT` peripheral is a hardware module available in `ESP`
-//! chips, which functions as a pulse counter and encoder. It is capable of
-//! counting pulses and monitoring changes in signal levels from external
-//! sources.
+//! The PCNT module is designed to count the number of rising
+//! and/or falling edges of input signals. The may contain multiple pulse
```
Comment (JurajSadel) [inline esp-hal/src/pcnt/mod.rs:5]:
```suggestion
//! and/or falling edges of input signals. They may contain multiple pulse
```
Did you mean something like this?

File: esp-hal/src/spi/mod.rs [hunk 7]
```diff
@@ -1,5 +1,10 @@
-//! Serial Peripheral Interface
+//! Serial Peripheral Interface (SPI)
 //!
+//! ## Overview
+//! The Serial Peripheral Interface (SPI) is a synchronous serial interface
+//! useful for communication with external peripherals.
```
Comment (JurajSadel) [inline esp-hal/src/spi/mod.rs:5]:
```suggestion
//! used for communication with external peripherals.
```

File: esp-hal/src/spi/slave.rs [hunk 8]
```diff
@@ -57,6 +50,15 @@
 //! transfer.wait().unwrap();
 //! # }
 //! ```
+//! 
+//! ## Implementation State
```
Comment (JurajSadel) [inline esp-hal/src/spi/slave.rs:54]: Maybe we can add here that ESP32 SPI Slave is not currently supported?

File: esp-hal/src/timer/timg.rs [hunk 9]
```diff
@@ -1,25 +1,26 @@
...
 //!
 //! The general-purpose timers are based on a 16-bit pre-scaler and a 54-bit
-//! auto-reload-capable up-down counter. The timers have configurable alarms,
-//! which are triggered when the internal counter of the timers reaches a
-//! specific target value. The timers are clocked using the APB clock source.
+//! auto-reload-capable up-down counter.
+//!
+//! ## Configuration
+//! The timers have configurable alarms, which are triggered when the internal
+//! counter of the timers reaches a  specific target value. The timers are
```
Comment (JurajSadel) [inline esp-hal/src/timer/timg.rs:12]:
```suggestion
//! counter of the timers reaches a specific target value. The timers are
```

File: esp-hal/src/twai/mod.rs [hunk 10]
```diff
@@ -1,22 +1,25 @@
...
 //!
 //! ## Overview
-//! The TWAI peripheral driver provides functions and structs specifically
-//! designed for the Two-Wire Automotive Interface (TWAI) on ESP chips. TWAI is
-//! a communication protocol commonly used in automotive applications for
-//! sending and receiving messages between electronic control units (ECUs) in a
-//! vehicle.
+//! The TWAI is a multi-master, multi-cast communication protocol with error
+//! detection and signaling and inbuilt message priorities and arbitration. The
+//! TWAI protocol is suited for automotive and industrial applications
```
Comment (JurajSadel) [inline esp-hal/src/twai/mod.rs:6]:
```suggestion
//! TWAI protocol is suited for automotive and industrial applications.
```

**Review by JurajSadel — COMMENTED** [review]
Thank you for working on this! LGTM, I left a few comments.

File: esp-hal/src/spi/mod.rs (hunk 7 above)
Comment (SergioGasquez) [reply esp-hal/src/spi/mod.rs:5]: This was taken from the TRM 
image

File: esp-hal/src/interrupt/xtensa.rs [hunk 11]
```diff
@@ -1,3 +1,5 @@
+//! Interrupt handling - Xtensa
```
Comment (jessebraham) [inline esp-hal/src/interrupt/xtensa.rs:1]:
```suggestion
//! Interrupt handling
```

I think we probably don't need to specify the arch here. We build the documentation for each chip, so each corresponding set of documentation should only include the relevant bits anyway.

File: esp-hal/src/dma/mod.rs [hunk 12]
```diff
@@ -1,20 +1,17 @@
...
 //!
 //! ## Overview
-//!
-//! The `DMA` driver provides an interface to efficiently transfer data between
+//! The DMA driver provides an interface to efficiently transfer data between
 //! different memory regions and peripherals within the ESP microcontroller
-//! without involving the CPU. The `Direct Memory Access` (DMA) controller is a
-//! hardware block responsible for managing these data transfers.
+//! without involving the CPU. The DMA controller is a hardware block
+//! responsible for managing these data transfers.
```
Comment (jessebraham) [inline esp-hal/src/dma/mod.rs:7]:
```suggestion
//! without involving the CPU. The DMA controller is responsible for
//! managing these data transfers.
```

This sounds a bit awkward to me, and can be simplified IMO.

**Review by jessebraham — CHANGES_REQUESTED** [review]
Couple quick comments, I'm a little over half done reviewing and looks quite good so far. Will try to wrap this up later today or tomorrow morning.

File: esp-hal/src/soc/esp32s2/gpio.rs [hunk 13]
```diff
@@ -67,11 +67,16 @@ pub const NUM_PINS: usize = 47;
 
 pub(crate) const FUNC_IN_SEL_OFFSET: usize = 0;
 
+#[doc(hidden)]
```
Comment (jessebraham) [inline esp-hal/src/soc/esp32s2/gpio.rs:70]: I think these types can be `pub(crate)` instead?

File: esp-hal/src/soc/esp32s3/gpio.rs [hunk 14]
```diff
@@ -55,11 +55,16 @@ pub const NUM_PINS: usize = 49;
 
 pub(crate) const FUNC_IN_SEL_OFFSET: usize = 0;
 
+#[doc(hidden)]
```
Comment (jessebraham) [inline esp-hal/src/soc/esp32s3/gpio.rs:58]: See above.

**Review by jessebraham — CHANGES_REQUESTED** [review]
Just a few more little things (sorry!) and then I think this is probably good to go.

**Review by jessebraham — APPROVED** [review]
Thanks, LGTM!

**Review by JurajSadel — APPROVED** [review]
LGTM, thanks!

**Pull request merged**